### PR TITLE
Nbdev2

### PR DESCRIPTION
**Pull request by steveri**
This is phase 1 of the master plan, see email for details.  These changes should suffice to make all our existing tests (8x8 pointwise, conv12, conv21, conv31, convbw) pass in travis using SMT-PNR along with the latest version of the CGRA generator (nbdev2).

### CODE REVIEW

File: makefile [hunk 1]
```diff
@@ -223,20 +239,21 @@ endif
...
-	cd CGRAGenerator; ./bin/generate.csh $(QVSWITCH) || exit -1
+	cd CGRAGenerator; ./bin/generate.csh $(QVSWITCH) || exit 13
 	cp CGRAGenerator/hardware/generator_z/top/cgra_info.txt build/cgra_info_4x4.txt
 
 build/cgra_info_8x8.txt:
 	@echo; echo Making $@ because of $?
 	@echo "CGRA generate (generates 8x8 CGRA + connection matrix for pnr)"
-	cd CGRAGenerator; export CGRA_GEN_USE_MEM=1; ./bin/generate.csh $(QVSWITCH) || exit -1
+	cd CGRAGenerator; export CGRA_GEN_USE_MEM=1; ./bin/generate.csh $(QVSWITCH) || exit 13
+#	cd CGRAGenerator; export CGRA_GEN_USE_MEM=1; ./bin/generate.csh -v || exit 13
```
Comment (leonardt) [inline makefile]: Need to keep this comment around?

File: makefile [hunk 2]
```diff
@@ -81,7 +97,7 @@ core_tests:
 	make build/conv_2_1.correct.txt  DELAY=10,0  GOLD=ignore
 	make build/conv_3_1.correct.txt  DELAY=20,0  GOLD=ignore
 	make build/conv_bw.correct.txt   DELAY=130,0 GOLD=ignore
-	make build/cascade_mapped.json GOLD=ignore
+#	make build/cascade_mapped.json GOLD=ignore
```
Comment (leonardt) [inline makefile]: Need to keep this comment around? Maybe add a comment as to why it's commented out (so we remember)

**Comment by rdaly525** [comment]
Because there are changes to the flow, the following should happen roughly at the same time:

-Do one final commit in this CGRAFlow branch changing the names of the P&R/CGRA branches back to 'master'.
-Merge this pull request.
-Merge P&R and CGRA branches into their own master.
-Verify that master for the flow did not break.

**Comment by steveri** [comment]
Re: "the following should happen roughly at the same time"

Actually I was thinking we could preserve a 'never-broken' master by doing the following:

1. check in this master, which is working, and which points to 'nbdev2' and 'io_tile' branches.

2. At their leisure (but presumably soon) Steve merges CGRAGenerator/nbdev2 to CGRAGenerator/master and Caleb or Makai merges SMT/io_tiles to SMT/master such that now

* CGRAGenerator/nbdev2 == CGRAGenerator/master and 
* SMT/io_tiles == SMT/master

3. Again at their leisure (but presumably soon), CGRAFlow is redirected to point at CGRAGenerator/master and SMT/master.

To that end, I will go ahead and merge this pull to get the ball rolling...

SR

**Comment by leonardt** [comment]
I would vote for rewording **at their leisure** to **ASAP**. I believe we already saw a potential issue related to stale master branches (issue with taeyoung's base branch). In terms of methodology, I think it's best practice to always keep master fresh. This way, a team member can branch off master with confidence that it's up to date, and they don't have to hunt down the right branch to base off of.

**Comment by leonardt** [comment]
I would also vote that we don't merge into CGRAFlow master with the travis script pointing to non-master branches (again to promote the idea that individual repositories are master is fresh)

**Comment by makaimann** [comment]
I agree that branches should be merged ASAP. PNR io_tiles branch is ready to be merged. In keeping with other good practices, I'm having @cdonovick review and commit my changes. I think he's busy today, but I anticipate that will be merged by EOD Tuesday. Regarding CGRAFlow master pointing to non-master branches, I agree it's probably better not to. I don't think it's worth reverting though.

**Comment by steveri** [comment]
Sorry, I guess I jumped the gun.   FWIW at this point I merged CGRAGenerator/nbdev2 and CGRAGenerator/master, and CGRAFlow is currently pointing to CGRAGenerator/master and passing (build 1209).  Looks like SMT branches will be sorted out soon, and maybe we can move forward from there.

In keeping with the will of the crowd, I will try and remember to order it differently next time around... :)

**Comment by steveri** [comment]
There's a third option, I guess, that may or may not make everyone happy.  It preserves the never-break rule for master and also the (new?) master-only-points-to-master-sub-repositories rule.  Next time we could do the following, unless people decide it's too complicated and/or prefer a different path:

1. Build a CGRAFlow/staging branch that points to the new sub-repository branches
2. merge all sub-repository branches to their respective master branches
3. Switch CGRAFlow/staging to point to master branches only
4. Merge CGRAFlow/staging to become the new CGRAFlow/master

**Comment by steveri** [comment]
PS On the previous comment, in case it wasn't obvious, in our current situation it would have been the case that 'staging' == 'nbdev2'.  Which I guess is what Ross was proposing all along, wasn't it :)

**Comment by leonardt** [comment]
That seems like a reasonable path. Should be possible to write a script that uses the GitHub API to perform those sequence of merges in order once the original (step 1) staging branch is green. I'd be willing to take that on once I get some time

**Comment by cdonovick** [comment]
P&R is merged.

**Comment by steveri** [comment]
Okay I changed CGRAFlow/master to point to ONLY MASTER BRANCHES, maybe now the world will be right again :)

**Comment by steveri** [comment]
master build passed, maybe we're finally done with this thread?

https://travis-ci.org/StanfordAHA/CGRAFlow/builds/338615543

**Pull request merged**